File: finance1summaryagent/server/crm_to_text.py
"""
Converts Zoho CRM record data to formatted text for LLM consumption.
"""
from typing import Dict, Any


def record_to_text(record_data: Dict[str, Any], entity_type: str = "Record") -> str:
    """
    Converts a Zoho CRM record dictionary to a formatted text string.
    
    Args:
        record_data: Record data dictionary from Zoho API
        entity_type: Module name (e.g., "Accounts", "Deals")
    
    Returns:
        Formatted text string representation of the record.
    """
    if not record_data:
        return f"{entity_type} record: No data available"
    
    lines = [f"{entity_type} Record Information:"]
    lines.append("=" * 50)
    
    # Extract ID if present
    record_id = record_data.get("id") or record_data.get("Id")
    if record_id:
        lines.append(f"ID: {record_id}")
    
    # Format all fields
    for key, value in record_data.items():
        # Skip internal fields
        if key.lower() in ["id", "created_time", "modified_time", "created_by", "modified_by"]:
            continue
        
        # Format the field name (replace underscores, capitalize)
        field_name = key.replace("_", " ").title()
        
        # Format the value
        if value is None:
            formatted_value = "Not set"
        elif isinstance(value, dict):
            # Handle nested objects (e.g., Owner, Account_Name)
            if "name" in value:
                formatted_value = value["name"]
            elif "id" in value:
                formatted_value = f"{value.get('name', 'Unknown')} (ID: {value['id']})"
            else:
                formatted_value = str(value)
        elif isinstance(value, list):
            formatted_value = ", ".join(str(v) for v in value)
        else:
            formatted_value = str(value)
        
        lines.append(f"{field_name}: {formatted_value}")
    
    return "\n".join(lines)

File: supabase/functions/_shared/crmToText.ts
/**
 * Convert Zoho CRM record to formatted text for LLM.
 * Port of server/crm_to_text.py
 */

export function recordToText(
  recordData: Record<string, unknown>,
  entityType: string = "Record"
): string {
  if (!recordData || Object.keys(recordData).length === 0) {
    return `${entityType} record: No data available`;
  }

  const lines: string[] = [`${entityType} Record Information:`, "=".repeat(50)];

  const recordId = recordData.id ?? recordData.Id;
  if (recordId != null) {
    lines.push(`ID: ${recordId}`);
  }

  const skipKeys = new Set(["id", "created_time", "modified_time", "created_by", "modified_by"]);

  for (const [key, value] of Object.entries(recordData)) {
    if (skipKeys.has(key.toLowerCase())) continue;

    const fieldName = key.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());

    let formattedValue: string;
    if (value == null) {
      formattedValue = "Not set";
    } else if (typeof value === "object" && !Array.isArray(value) && value !== null) {
      const obj = value as Record<string, unknown>;
      if ("name" in obj) {
        formattedValue = String(obj.name);
      } else if ("id" in obj) {
        formattedValue = `${obj.name ?? "Unknown"} (ID: ${obj.id})`;
      } else {
        formattedValue = JSON.stringify(value);
      }
    } else if (Array.isArray(value)) {
      formattedValue = value.map((v) => String(v)).join(", ");
    } else {
      formattedValue = String(value);
    }

    lines.push(`${fieldName}: ${formattedValue}`);
  }

  return lines.join("\n");
}
